pkg/storage/ingest/pusher.go

```diff
@@ -241,163 +253,239 @@ func (c pusherConsumer) shouldLogClientError(ctx context.Context, err error) (bo
-	c.pushers[userID] = p
-	return p
+
+	shards := c.shardsFor(userID)
+	return shards.ShardWriteRequest(ctx, wr)
 }
 
-func (c multiTenantPusher) Close() []error {
+// Close implements the PusherCloser interface.
+func (c parallelStoragePusher) Close() []error {
 	var errs multierror.MultiError
 	for _, p := range c.pushers {
-		errs.Add(p.close())
+		errs.Add(p.Stop())
 	}
 	clear(c.pushers)
 	return errs
 }
 
-type shardingPusher struct {
-	numShards             int
-	shards                []chan flushableWriteRequest
-	unfilledShards        []flushableWriteRequest
-	upstream              Pusher
-	wg                    *sync.WaitGroup
-	errs                  chan error
-	batchSize             int
+func (c parallelStoragePusher) shardsFor(userID string) *parallelStorageShards {
+	if p := c.pushers[userID]; p != nil {
+		return p
+	}
+	p := newParallelStorageShards(c.metrics.numTimeSeriesPerFlush, c.numShards, c.batchSize, shardForSeriesBuffer, c.upstreamPusher)
+	c.pushers[userID] = p
+	return p
+}
+
+type parallelStorageShards struct {
 	numTimeSeriesPerFlush prometheus.Histogram
+
+	pusher Pusher
+
+	numShards int
+	batchSize int
+	capacity  int
+
+	wg     *sync.WaitGroup
+	shards []*BatchingQueue
+	errs   chan error
+}
+
+type FlushableWriteRequest struct {
+	*mimirpb.WriteRequest
+	context.Context
 }
 
 // TODO dimitarvdimitrov if this is expensive, consider having this long-lived and not Close()-ing and recreating it on every fetch, but instead calling something like Flush() on it.
-func newShardingPusher(numTimeSeriesPerFlush prometheus.Histogram, numShards int, batchSize int, buffer int, upstream Pusher) *shardingPusher {
-	pusher := &shardingPusher{
+func newParallelStorageShards(numTimeSeriesPerFlush prometheus.Histogram, numShards int, batchSize int, capacity int, pusher Pusher) *parallelStorageShards {
+	p := &parallelStorageShards{
 		numShards:             numShards,
-		upstream:              upstream,
+		pusher:                pusher,
+		capacity:              capacity,
 		numTimeSeriesPerFlush: numTimeSeriesPerFlush,
 		batchSize:             batchSize,
 		wg:                    &sync.WaitGroup{},
 		errs:                  make(chan error, numShards),
-		unfilledShards:        make([]flushableWriteRequest, numShards),
-	}
-	shards := make([]chan flushableWriteRequest, numShards)
-	pusher.wg.Add(numShards)
-	for i := range shards {
-		pusher.unfilledShards[i].WriteRequest = &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()}
-		shards[i] = make(chan flushableWriteRequest, buffer)
-		go pusher.runShard(shards[i])
 	}
-	go func() {
-		pusher.wg.Wait()
-		close(pusher.errs)
-	}()
 
-	pusher.shards = shards
-	return pusher
+	p.start()
+
+	return p
 }
 
-func (p *shardingPusher) PushToStorage(ctx context.Context, request *mimirpb.WriteRequest) error {
+func (p *parallelStorageShards) ShardWriteRequest(ctx context.Context, request *mimirpb.WriteRequest) error {
```

Comment:
I introduced different names here as I found it really strange that we had levels of equal method name that are not part of interface e.g. `shardingPusher` was not fulfilled the `PusherCloser` interface but had similar method names.